### PR TITLE
Fix unpacking process for the compressed folders 

### PR DESCRIPTION
**Pull request by usepowershell**
Currently, when files are updated locally and a new kitchen action is taken, the files on the remote node are not updated.  This PR changes the unpacking and relocation of the files to make sure that new and updated files are moved into the proper locations.

I've also started adding Pester tests to the supporting PowerShell functions.

### CODE REVIEW

**Comment by usepowershell** [comment]
Looks like there are some embedded base64 strings in the tests.  I'll have to update those.

**Comment by tyler-ball** [comment]
I have no idea what I am looking at.  @fnichol and @jmundrawala and @adamex?

**Comment by fnichol** [comment]
@smurawski hopefully #5 (merged) will help with the failure you were seeing today.

**Comment by usepowershell** [comment]
@fnichol @tyler-ball The tests should be fixed, but it should be noted that they are very whitespace dependent.

**Comment by usepowershell** [comment]
The build is failing due to style guidelines.  I'll have to poke at it next week.

**Comment by usepowershell** [comment]
@tyler-ball @fnichol Build is good. Quality checks ok as well now.

**Comment by tyler-ball** [comment]
I'm :+1: but I'm useless for the Powershell.  @fnichol and @jdmundrawala?

File: support/decode_files.ps1 [hunk 1]
```diff
@@ -1,60 +1,50 @@
...
-    $tmp = Unresolve-Path $_.Key
+  foreach ($key in $hash.keys) {
+    $value = $hash[$key]
+    $tmp, $tzip, $dst = $Key, $Value["tmpzip"], $Value["dst"]
     $sMd5 = (Get-Item $tmp).BaseName.Replace("b64-", "")
-    $tzip, $dst = (Unresolve-Path $_.Value["tmpzip"]), (Unresolve-Path $_.Value["dst"])
     $decoded = if ($tzip -ne $null) { $tzip } else { $dst }
     Decode-Base64File $tmp $decoded
-    Remove-Item $tmp -Force
+    #Remove-Item $tmp -Force
```
Comment (fnichol) [inline support/decode_files.ps1]: Can this line be removed? Looks like a comment…

File: lib/winrm/transport/file_transporter.rb [hunk 2]
```diff
@@ -253,7 +252,10 @@ def decode_files(files)
           vars = %{$hash_file = "#{hash_file}"\n}
 
           output = service.run_powershell_script(
-            [vars, decode_files_script].join("\n")
+            [vars,
+             ". #{decode_files_script}",
+             "Decode-Files (Invoke-Input $hash_file) | ConvertTo-Csv -NoTypeInformation"
```
Comment (fnichol) [inline lib/winrm/transport/file_transporter.rb:257]: Not a bad idea, I take it this make the functions much easier to test via Pester?

File: support/decode_files.ps1 [hunk 3]
```diff
@@ -1,60 +1,50 @@
...
-}
-
+Function Cleanup($o) { if (($o) -and ($o.GetType().GetMethod("Dispose") -ne $null)) { $o.Dispose() } }
+Function Decode-Base64File($src, $dst) {set-content -Encoding Byte -Path $dst -Value ([Convert]::FromBase64String([IO.File]::ReadAllLines($src)))}
 Function Copy-Stream($src, $dst) { $b = New-Object Byte[] 4096; while (($i = $src.Read($b, 0, $b.Length)) -ne 0) { $dst.Write($b, 0, $i) } }
+function Resolve-ProviderPath{ $input | % {if ($_){(Resolve-Path $_).ProviderPath} else{$null}} }
+Function Release-COM($o) { if ($o -ne $null) { [void][Runtime.Interopservices.Marshal]::ReleaseComObject($o) } }
+function Test-NETStack($Version, $r = 'HKLM:\Software\Microsoft\NET Framework Setup\NDP\v4') { [bool]("$r\Full", "$r\Client" | ? {(gp $_).Version -like "$($Version)*"}) }
+function Test-IOCompression {($PSVersionTable.PSVersion.Major -ge 3) -and (Test-NETStack '4.5')}
+set-alias RPP -Value Resolve-ProviderPath
```
Comment (fnichol) [inline support/decode_files.ps1:8]: Alias, good idea!

File: support/decode_files.ps1 [hunk 4]
```diff
@@ -1,60 +1,50 @@
...
 
 Function Get-MD5Sum($src) {
   Try {
-    $c = New-Object -TypeName System.Security.Cryptography.MD5CryptoServiceProvider
+    $c = New-Object -TypeName Security.Cryptography.MD5CryptoServiceProvider
     $bytes = $c.ComputeHash(($in = (Get-Item $src).OpenRead()))
-    return ([System.BitConverter]::ToString($bytes)).Replace("-", "").ToLower()
-  } Finally { Cleanup $c; Cleanup $in }
+    ([BitConverter]::ToString($bytes)).Replace("-", "").ToLower()
+  } catch [exception]{throw $_} finally { Cleanup $c; Cleanup $in }
```
Comment (fnichol) [inline support/decode_files.ps1:30]: This seems much safer now that I understand PowerShell doesn't re-raise the exception implicitly.

File: support/decode_files.ps1 [hunk 5]
```diff
@@ -1,60 +1,50 @@
...
 }
 
 Function Invoke-Input($in) {
-  $in = Unresolve-Path $in
+  $in = $in | rpp
   Decode-Base64File $in ($decoded = "$($in).ps1")
   $expr = Get-Content $decoded | Out-String
   Remove-Item $in,$decoded -Force
-  return Invoke-Expression "$expr"
+  Invoke-Expression "$expr"
```
Comment (fnichol) [inline support/decode_files.ps1:38]: Would `Invoke-Expression` ever return the result of the expression? That was a guess on my part, hence the return in the original code. I yield to your experience :smile:

**Comment by fnichol** [comment]
@smurawski Aside from potentially deleting a few more bytes in the commented line, I'm :+1: for merge

File: lib/winrm/transport/file_transporter.rb (hunk 2 above)
Comment (usepowershell) [inline lib/winrm/transport/file_transporter.rb:257]: Yeah, when we ran the Decode-Files line in the script it makes testing a bit tougher.

File: support/decode_files.ps1 (hunk 1 above)
Comment (usepowershell) [inline support/decode_files.ps1]: oops, I think that should be uncommented.. that was a troubleshooting step.  I'll fix that.

File: support/decode_files.ps1 (hunk 3 above)
Comment (usepowershell) [inline support/decode_files.ps1:8]: It saves a few characters, since it's called a few times.

File: support/decode_files.ps1 (hunk 1 above)
Comment (usepowershell) [inline support/decode_files.ps1]: Change made

**Pull request merged**